aws/logs_monitoring/lambda_function.py

```diff
@@ -591,7 +584,23 @@ def s3_handler(event, context, metadata):
         with gzip.GzipFile(fileobj=BytesIO(data)) as decompress_stream:
             # Reading line by line avoid a bug where gzip would take a very long time (>5min) for
             # file around 60MB gzipped
-            data = "".join(BufferedReader(decompress_stream))
+
+            # Only make another copy of our log data if we're performing some operation
+            # against the entire block of data. Need to conserve memory in lambda land.
+            if DD_MULTILINE_LOG_REGEX_PATTERN or is_cloudtrail(str(key)):
+                data = "".join(BufferedReader(decompress_stream))
+            else:
+                # Send decompressed lines to Datadog. Stream iter is line-by-line already.
+
+                # NOTE: Duplicates the non-compressed case below, but we can't allow falling through to that
+                #   since we'll lose the decompress_stream context.
+                for line in decompress_stream:
```

Comment:
this block is duplicate from https://github.com/DataDog/datadog-serverless-functions/pull/117/files#diff-618731ddf1c446ef45d09613cb4b189bR622, do you think we can factorise this?

Reply:
Good call.  Yes, I'll push through an update shortly.